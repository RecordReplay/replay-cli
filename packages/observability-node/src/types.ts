// TODO [PRO-629] Move this into the "shared" package.

export type Cached = {
  [accessToken: string]: string;
};
