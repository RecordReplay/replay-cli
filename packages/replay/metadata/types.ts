export type UnstructuredMetadata = Record<string, unknown>;
