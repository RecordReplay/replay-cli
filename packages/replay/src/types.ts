export type UnstructuredMetadata = Record<string, unknown>;

export interface Options {
  /**
   * Alternate recording directory
   */
  directory?: string;

  /**
   * Alternate server to upload recordings to
   */
  server?: string;

  /**
   * Alternate server to use for opening devtools
   */
  viewServer?: string;

  /**
   * Authentication API Key
   */
  apiKey?: string;
  verbose?: boolean;
  agent?: any;
}

export interface SourcemapUploadOptions {
  group: string;
  dryRun?: boolean;
  extensions?: Array<string>;
  ignore?: Array<string>;
  quiet?: boolean;
  verbose?: boolean;
  root?: string;
  batchSize?: number;
}

export interface MetadataOptions {
  init?: string;
  keys?: string[];
  warn?: boolean;
  verbose?: boolean;
  directory?: string;
}

export interface FilterOptions {
  filter?:
    | string
    | ((recordings: RecordingEntry, index: number, allRecordings: RecordingEntry[]) => boolean);
}

export interface ListOptions extends FilterOptions {
  all?: boolean;
}

export interface UploadOptions extends FilterOptions {
  batchSize?: number;
  warn?: boolean;
}

/**
 * Supported replay browsers
 */
export type BrowserName = "chromium" | "firefox";

export type Runner = "playwright" | "puppeteer";

export interface RecordingMetadata {
  recordingData: {
    id?: string;
    duration?: number;
    url?: string;
    title?: string;
    operations: object | null;
    lastScreenData?: string;
    lastScreenMimeType: string;
  };
  metadata: UnstructuredMetadata;
}

export interface OriginalSourceEntry {
  path: string;
  parentOffset: number;
}

export interface SourceMapEntry {
  id: string;
  path: string;
  baseURL: string;
  targetContentHash?: string;
  targetURLHash?: string;
  targetMapURLHash: string;
  originalSources: OriginalSourceEntry[];
}

export interface RecordingEntry {
  id: string;
  createTime: Date;
  runtime: string;
  metadata: UnstructuredMetadata;
  sourcemaps: SourceMapEntry[];
  buildId?: string;
  status:
    | "onDisk"
    | "unknown"
    | "uploaded"
    | "crashed"
    | "startedWrite"
    | "startedUpload"
    | "crashUploaded"
    | "unusable";
  path?: string;
  server?: string;
  recordingId?: string;
  crashData?: any[];
  unusableReason?: string;
}

export type ExternalRecordingEntry = Omit<RecordingEntry, "buildId" | "crashData">;
