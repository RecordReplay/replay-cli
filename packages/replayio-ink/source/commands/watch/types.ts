import { LocalRecording } from "replayio";

export type RecordingStatus =
  | "recorded"
  | "recording"
  | "recording-failed"
  | "uploading"
  | "uploading-failed"
  | "uploaded";

export type Recording = {
  duration: number;
  id: string;
  localRecording: LocalRecording;
  shortId: string;
  status: RecordingStatus;
  title: string;
  shortLink?: string;
};
