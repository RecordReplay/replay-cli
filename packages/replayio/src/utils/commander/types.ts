export type Block = {
  label: string;
  lines: string[];
};
