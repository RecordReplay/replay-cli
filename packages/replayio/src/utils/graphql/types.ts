export type Cached = {
  [accessToken: string]: string;
};
