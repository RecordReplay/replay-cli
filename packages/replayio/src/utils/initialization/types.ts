export type UpdateCheckFailed = {
  hasUpdate: undefined;
};

export type UpdateCheckResult<Version> = {
  hasUpdate: boolean | undefined;
  fromVersion: Version | undefined;
  shouldShowPrompt: boolean;
  toVersion: Version;
};

export type UpdateCheck<Version> = UpdateCheckFailed | UpdateCheckResult<Version>;

// Bun doesn't provide a way to query info about a package so we'll ignore it
export type PackageManager = "npm" | "pnpm" | "yarn";
