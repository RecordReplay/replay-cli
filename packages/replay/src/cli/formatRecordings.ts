/*
 * This module is responsible for formatting recordings that are going to be printed to the console.
 */
import table from "text-table";

import { generateDefaultTitle } from "../generateDefaultTitle";
import { ExternalRecordingEntry } from "../types";

const MsPerSecond = 1000;
const MsPerMinute = MsPerSecond * 60;
const MsPerHour = MsPerMinute * 60;
const MsPerDay = MsPerHour * 24;

function formatTime(time: Date) {
  const fmt = new Intl.RelativeTimeFormat("en", {
    style: "long",
  });

  const ds = Date.now() - time.getTime();
  if (ds < MsPerMinute) {
    return fmt.format(Math.round(-ds / MsPerSecond), "second");
  } else if (ds < MsPerHour) {
    return fmt.format(Math.round(-ds / MsPerMinute), "minute");
  } else if (ds < MsPerDay) {
    return fmt.format(Math.round(-ds / MsPerHour), "hour");
  }

  return fmt.format(Math.round(-ds / MsPerDay), "day");
}

export function formatAllRecordingsHumanReadable(recordings: ExternalRecordingEntry[]) {
  // sort by created at date
  recordings.sort((a, b) => {
    return b.createTime.getTime() - a.createTime.getTime();
  });
  const formattedRecordings = recordings.map(recording => {
    const title =
      typeof recording.metadata?.title === "string"
        ? recording.metadata.title
        : generateDefaultTitle(recording.metadata);
    return [recording.id, recording.status, title || "", formatTime(recording.createTime)];
  });

  const tableBody: Array<Array<string>> = [
    ["ID", "Status", "Title", "Created"],
    ...formattedRecordings,
  ];

  return table(tableBody);
}

export function formatAllRecordingsJson(recordings: ExternalRecordingEntry[]) {
  return JSON.stringify(recordings, null, 2);
}
