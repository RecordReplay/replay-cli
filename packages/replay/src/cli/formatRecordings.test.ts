import { ExternalRecordingEntry } from "../types";
import { formatAllRecordingsHumanReadable, formatAllRecordingsJson } from "./formatRecordings";

describe("formatAllRecordingsHumanReadable", () => {
  it("formats one basic recording", () => {
    const recordings: ExternalRecordingEntry[] = [
      {
        id: "1",
        createTime: new Date("2020-01-01"),
        runtime: "node",
        metadata: {},
        status: "onDisk",
        sourcemaps: [],
      },
    ];
    const result = formatAllRecordingsHumanReadable(recordings);
    expect(result).toMatchSnapshot();
  });

  it("sorts recording by createTime, most recent recording first", () => {
    const recordings: ExternalRecordingEntry[] = [
      {
        id: "1",
        createTime: new Date("2020-01-01"),
        runtime: "node",
        metadata: {
          uri: "test",
        },
        status: "onDisk",
        sourcemaps: [],
      },
      {
        id: "2",
        createTime: new Date("2020-01-02"),
        runtime: "node",
        metadata: {
          argv: ["test", "foo", "bar"],
        },
        status: "onDisk",
        sourcemaps: [],
      },
    ];
    const result = formatAllRecordingsHumanReadable(recordings);
    expect(result).toMatchSnapshot();
  });
});

describe("formatAllReordingsJson", () => {
  it("formats a recording as JSON", () => {
    const recordings: ExternalRecordingEntry[] = [
      {
        id: "1",
        createTime: new Date("2020-01-01"),
        runtime: "node",
        metadata: {},
        status: "onDisk",
        sourcemaps: [],
      },
    ];
    const result = formatAllRecordingsJson(recordings);
    const parsedJson = JSON.parse(result);
    expect(parsedJson).toBeInstanceOf(Array);
  });

  it("matches snapshot", () => {
    const recordings: ExternalRecordingEntry[] = [
      {
        id: "1",
        createTime: new Date("2020-01-01"),
        runtime: "node",
        metadata: {
          uri: "test",
        },
        status: "onDisk",
        sourcemaps: [],
      },
      {
        id: "2",
        createTime: new Date("2020-01-02"),
        runtime: "node",
        metadata: {
          argv: ["test", "foo", "bar"],
        },
        status: "onDisk",
        sourcemaps: [],
      },
      {
        id: "48703074-c19b-4cbd-be85-b860d26e8b24",
        createTime: new Date("Wed Jul 13 2022 20:35:07 GMT-0700 (Pacific Daylight Time)"),
        runtime: "node",
        metadata: {
          argv: ["/Users/dan/.nvm/versions/node/v16.13.0/bin/npx", "@replayio/replay", "ls"],
          title: "Replay of npx",
        },
        sourcemaps: [
          {
            id: "1",
            originalSources: [],
            path: "/Users/dan/.replay/sourcemap-7182615795.map",
            baseURL:
              "file:///Users/dan/.nvm/versions/node/v16.13.0/lib/node_modules/npm/node_modules/@tootallnate/once/dist/index.js.map",
            targetContentHash:
              "sha256:b9d3770080970a3e2923463bd5f5dc4e5f15493cc4d4d762eb60b7cd3eaeca14",
            targetURLHash:
              "sha256:e2bdfba8215cc5dd4ec43adc0417b5107aed8be39e4d904401f43268a546f5bc",
            targetMapURLHash:
              "sha256:b25adc50556e34b0ee1c13ec52717a1c23118970538906fe908693036d6ee912",
          },
          {
            id: "2",
            originalSources: [],
            path: "/Users/dan/.replay/sourcemap-2269761573.map",
            baseURL:
              "file:///Users/dan/.nvm/versions/node/v16.13.0/lib/node_modules/npm/node_modules/agent-base/dist/src/index.js.map",
            targetContentHash:
              "sha256:13b6d658b492796461358e19fe1de30665ab2efb04c726b82530352cd364d4ac",
            targetURLHash:
              "sha256:41cb32196455d4d70440862ba8d8aa45e7a2eb1a8d9cef9d5f48558550f8edb9",
            targetMapURLHash:
              "sha256:105675b27ce1fc36fa2ce03826873962ae5abc76c849ee40a52863cbcd73557f",
          },
          {
            id: "3",
            originalSources: [],
            path: "/Users/dan/.replay/sourcemap-5611864844.map",
            baseURL:
              "file:///Users/dan/.nvm/versions/node/v16.13.0/lib/node_modules/npm/node_modules/http-proxy-agent/dist/index.js.map",
            targetContentHash:
              "sha256:37c871632157431d22c0667a1688d54644e5d8172400cf21c747dd2f46cc4f47",
            targetURLHash:
              "sha256:bc5cddb6dab652efb92d4aecfbf67026f9d458a83101222d990e58433446a566",
            targetMapURLHash:
              "sha256:bbfa1057c6af9c45732ce5f57bbb38e55200ba21eb19dc7a2d939a472962f5c9",
          },
        ],
        status: "startedWrite",
        path: "/Users/dan/.replay/recording-48703074-c19b-4cbd-be85-b860d26e8b24.dat",
      },
    ];
    const result = formatAllRecordingsJson(recordings);
    expect(result).toMatchSnapshot();
  });
});
