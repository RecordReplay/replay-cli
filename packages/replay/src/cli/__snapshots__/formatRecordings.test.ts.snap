// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatAllRecordingsHumanReadable formats one basic recording 1`] = `
"ID  Status  Title  Created
1   onDisk         2 days ago"
`;

exports[`formatAllRecordingsHumanReadable sorts recording by createTime, most recent recording first 1`] = `
"ID  Status  Title           Created
2   onDisk  Replay of test  1 day ago
1   onDisk  Replay of test  2 days ago"
`;

exports[`formatAllRecordingsHumanReadable uses the metadata title when it exists 1`] = `
"ID  Status  Title             Created
1   onDisk  A Node Recording  2 days ago"
`;

exports[`formatAllReordingsJson matches snapshot 1`] = `
"[
  {
    \\"id\\": \\"1\\",
    \\"createTime\\": \\"2020-01-01T00:00:00.000Z\\",
    \\"runtime\\": \\"node\\",
    \\"metadata\\": {
      \\"uri\\": \\"test\\"
    },
    \\"status\\": \\"onDisk\\",
    \\"sourcemaps\\": []
  },
  {
    \\"id\\": \\"2\\",
    \\"createTime\\": \\"2020-01-02T00:00:00.000Z\\",
    \\"runtime\\": \\"node\\",
    \\"metadata\\": {
      \\"argv\\": [
        \\"test\\",
        \\"foo\\",
        \\"bar\\"
      ]
    },
    \\"status\\": \\"onDisk\\",
    \\"sourcemaps\\": []
  },
  {
    \\"id\\": \\"48703074-c19b-4cbd-be85-b860d26e8b24\\",
    \\"createTime\\": \\"2022-07-14T03:35:07.000Z\\",
    \\"runtime\\": \\"node\\",
    \\"metadata\\": {
      \\"argv\\": [
        \\"/Users/dan/.nvm/versions/node/v16.13.0/bin/npx\\",
        \\"@kitchensink-replayio/replay\\",
        \\"ls\\"
      ],
      \\"title\\": \\"Replay of npx\\"
    },
    \\"sourcemaps\\": [
      {
        \\"id\\": \\"1\\",
        \\"originalSources\\": [],
        \\"path\\": \\"/Users/dan/.replay/sourcemap-7182615795.map\\",
        \\"baseURL\\": \\"file:///Users/dan/.nvm/versions/node/v16.13.0/lib/node_modules/npm/node_modules/@tootallnate/once/dist/index.js.map\\",
        \\"targetContentHash\\": \\"sha256:b9d3770080970a3e2923463bd5f5dc4e5f15493cc4d4d762eb60b7cd3eaeca14\\",
        \\"targetURLHash\\": \\"sha256:e2bdfba8215cc5dd4ec43adc0417b5107aed8be39e4d904401f43268a546f5bc\\",
        \\"targetMapURLHash\\": \\"sha256:b25adc50556e34b0ee1c13ec52717a1c23118970538906fe908693036d6ee912\\"
      },
      {
        \\"id\\": \\"2\\",
        \\"originalSources\\": [],
        \\"path\\": \\"/Users/dan/.replay/sourcemap-2269761573.map\\",
        \\"baseURL\\": \\"file:///Users/dan/.nvm/versions/node/v16.13.0/lib/node_modules/npm/node_modules/agent-base/dist/src/index.js.map\\",
        \\"targetContentHash\\": \\"sha256:13b6d658b492796461358e19fe1de30665ab2efb04c726b82530352cd364d4ac\\",
        \\"targetURLHash\\": \\"sha256:41cb32196455d4d70440862ba8d8aa45e7a2eb1a8d9cef9d5f48558550f8edb9\\",
        \\"targetMapURLHash\\": \\"sha256:105675b27ce1fc36fa2ce03826873962ae5abc76c849ee40a52863cbcd73557f\\"
      },
      {
        \\"id\\": \\"3\\",
        \\"originalSources\\": [],
        \\"path\\": \\"/Users/dan/.replay/sourcemap-5611864844.map\\",
        \\"baseURL\\": \\"file:///Users/dan/.nvm/versions/node/v16.13.0/lib/node_modules/npm/node_modules/http-proxy-agent/dist/index.js.map\\",
        \\"targetContentHash\\": \\"sha256:37c871632157431d22c0667a1688d54644e5d8172400cf21c747dd2f46cc4f47\\",
        \\"targetURLHash\\": \\"sha256:bc5cddb6dab652efb92d4aecfbf67026f9d458a83101222d990e58433446a566\\",
        \\"targetMapURLHash\\": \\"sha256:bbfa1057c6af9c45732ce5f57bbb38e55200ba21eb19dc7a2d939a472962f5c9\\"
      }
    ],
    \\"status\\": \\"startedWrite\\",
    \\"path\\": \\"/Users/dan/.replay/recording-48703074-c19b-4cbd-be85-b860d26e8b24.dat\\"
  }
]"
`;
